src/adagrams.js

```diff
@@ -1,15 +1,331 @@
+  let letterArray = Object.keys(LETTER_POOL);
+  let tempLetters = {};
+  let lettersInHand = [];
+
+  while (lettersInHand.length < 10) {
+    let newNum = Math.floor(Math.random() * (26 + 1));
+    let currentLetter = letterArray[newNum];
+    if (currentLetter === undefined) {
+      continue;
+    }
+    else if (!(currentLetter in tempLetters)) {
+      tempLetters[currentLetter] = 1;
+      lettersInHand.push(currentLetter);
+    }
+    else if (currentLetter in tempLetters) {
+      if (tempLetters[currentLetter] < LETTER_POOL[currentLetter]) {
+        tempLetters[currentLetter] += 1;
+        lettersInHand.push(currentLetter);
+      }
+    }
+    else {
+      continue;
+    }
+  }
+  return lettersInHand;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  let tempLetters = getNewObject(lettersInHand);
+  let inputObject = getNewObject(input);
+  for (let i = 0; i < input.length; i++) {
+    let currentLetter = input[i];
+    if (currentLetter in tempLetters) {
+      if (inputObject[currentLetter] <= tempLetters[currentLetter]) {
+        continue;
+      }
+      else {
+        return false;
+      }
+    }
+    else {
+      return false;
+    }
+  }
+  return true;
+};
+
+const getNewObject = (letters) => {
+  let newObject = {}
+  for (let i = 0; i < letters.length; i++) {
+    let currentLetter = letters[i];
+    if (!(currentLetter in newObject)) {
+      newObject[currentLetter] = 1;
+    }
+    else {
+      newObject[currentLetter] += 1;
+    }
+  }
+  return newObject;
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  let score = 0;
+  let capsWord = word.toUpperCase();
+  
+  for (let i = 0; i < word.length; i++) {
+    if (word === "") {
+      return 0;
+    }
+    let currentLetter = capsWord[i];
+    score += SCORE_BOARD[currentLetter];
+  }
+  if (word.length >= 7) {
+    score += 8;
+  }
+  return score;
 };
 
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  let scoreArray = getScoreArray(words);
+  let highScore = getHighScore(scoreArray);
+  let potentialWinners = getPotentialWinners(words, scoreArray, highScore);
+  let tenLetterWinner = checkForTen(potentialWinners);
+  if (tenLetterWinner) {
+    return {"word": tenLetterWinner, "score": highScore};
+  }
+  else {
+    let lengthArray = getLengthArray(potentialWinners);
+    let tieBreaker = getTieBreaker(potentialWinners, lengthArray);
+    return {"word": tieBreaker, "score": highScore};
+  }
+};
+
+const getScoreArray = (words) => {
+  let scoreArray = []
+  for (let i = 0; i < words.length; i++) {
+    let currentWord = words[i];
+    let currentScore = scoreWord(currentWord);
+    scoreArray.push(currentScore);
+  }
+  return scoreArray;
+};
+
+const getHighScore = (scoreArray) => {
+  let highScore = Math.max(...scoreArray);
+  return highScore;
+};
+
+const getPotentialWinners = (words, scoreArray, highScore) => {
```

Comment:
This method relies on the words array and the scoreArray to be the same length since they both use `i` to access the values. However relying on this fact and that the words and scores are in the same order can be tricky. If you have two lists holding words and their scores, could you use an object to associate words with their scores instead?

You can still iterate over the object to find the potentialWinners